Benchmark-user-guide (#36)
aai-institute/nnbench: docs/guides/benchmarks.md

# Benchmarks

You benchmark your machine learning code in nnbench by encapsulating the benchmarking code in separate functions and applying one of the decorators provided. 
The available decorators are 
- `@benchmark`, which runs a benchmark with supplied parameters,
- `@parametrize`, which runs several benchmarks with the supplied parameter configurations,

> - `@product`, which runs benchmarks with all parameter combinations that arise from the supplied values. 

Reviewer: We might want to give all of them as absolute imports like `@nnbench.benchmark`, since that's what we have done so far in tests and examples.

We will first introduce a small machine learning example which we will subsequently use to showcase the motivation for and usage of the benchmarking decorators.

> We recommend to compartmentalize the logic of the machine learning model, the benchmarks, and benchmark execution into different files. In this guide, these are called `training.py`, `benchmarks.py`, and `run_benchmarks.py`.

Reviewer: ```suggestion
We recommend to split the model training, benchmark definition, and benchmark running into different files. In this guide, these are called `training.py`, `benchmarks.py`, and `main.py`.
```

Then adjust all occurrences of `run_benchmarks.py` into main.py.

## Example
Let us consider an example where we want to evaluate a `scikit-learn`'s `RandomForestClassifier` on the "Iris" example dataset.
For this purpose, we will define several helper functions inside a file, `training.py`. We use `prepare_data()`, to load the dataset,  `train_rf()` to train a random forest model with the specified parameters, and `accuracy()` to calculate the accuracy of the supplied model on the given dataset.

```python
# training.py
import numpy as np
from sklearn import base, metrics
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

> def prepare_data() -> tuple[np.ndarray, np.ndarray,np.ndarray, np.ndarray]:

Reviewer: ```suggestion
def prepare_data() -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
```

I guess this is not linted by ruff?

    data = load_iris()
    X, y = data.data, data.target
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
    return X_train, X_test, y_train, y_test

def train_rf(X_train: np.ndarray, y_train: np.ndarray, n_estimators: int, max_depth: int, random_state: int = 42) -> RandomForestClassifier:
    model = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=random_state)
    model.fit(X_train, y_train)
    return model

def accuracy(model: base.BaseEstimator, y_test: np.ndarray, y_pred: np.ndarray) -> float:
    accuracy = metrics.accuracy_score(y_test, y_pred)
    return accuracy
```

## `@benchmark` for single benchmarks

> Now, inside a new file called `benchmarks.py` we define our benchmarks.

Reviewer: ```suggestion
Now, we define our benchmarks in a new file called `benchmarks.py`.
```

> We first encapsulate the benchmark logic into a function, `benchmark_accuracy()` which prepares the data, trains a classifier, and lastly, obtains the accuracy.

Reviewer: I mean, if we're doing this, we might as well make it the whole pipeline. Realistically, model training and preparation should not be a part of the function.

To mark such a function as a benchmark, we apply the `@benchmark` decorator.

```python
# benchmarks.py
import nnbench
from training import prepare_data, train_rf, accuracy

@nnbench.benchmark()
def benchmark_accuracy(n_estimators: int, max_depth: int, random_state: int) -> float:
    X_train, X_test, y_train, y_test = prepare_data()
    rf = train_rf(X_train=X_train, y_train=y_train, n_estimators=n_estimators,
                  max_depth=max_depth, random_state=random_state)
    acc = accuracy(model=rf, X_test=X_test, y_test=y_test)
    return acc
```

Lastly, we set up a benchmark runner in the `run_benchmarks.py`. There, we supply the parameters (`n_estimators`, `max_depth`, `random_state`) necessary in the function definition. 

```python
# run_benchmarks.py
from nnbench import runner

r = runner.BenchmarkRunner()
result = r.run("./benchmarks.py", params={"n_estimators": 100, "max_depth": 5, "random_state": 42})
r.report(to='console', result=result)
```

When we execute the `run_benchmarks.py` we get the following output.

```bash
python run_benchmarks.py  

name         value
--------  --------
accuracy  0.933333
```

> # `@parametrize` for multiple configuration benchmarks

Reviewer: ```suggestion
## `@nnbench.parametrize` for multiple configuration benchmarks
```

(Please use double ## for subheadings.)

Sometimes, we are not only interested in the performance of a model for given parameters but want to compare the performance for different configurations. 

> To achieve this, let's modify our `benchmarks.py` to run a parametrized benchmark. 

Reviewer: ```suggestion
To achieve this, we can turn our single accuracy benchmark in the `benchmarks.py` file into a parametrized benchmark.
```

> We need to replace the decorator with our `@parametrize` configuration and supply the wanted parameter combinations within an iterable.

Reviewer: ```suggestion
To do this, replace the decorator with `@nnbench.parametrize` and supply the parameter combinations of choice as dictionaries in the first argument.
```

```python
# benchmarks.py
import nnbench
from training import prepare_data, train_rf, accuracy

@nnbench.parametrize((
    {"n_estimators": 10, "max_depth": 2},
    {"n_estimators": 50, "max_depth": 5},
    {"n_estimators": 100, "max_depth": 10}
))
def benchmark_accuracy(n_estimators: int, max_depth: int, random_state: int) -> float:

>     ... #   Our benchmarking logic from above

Reviewer: Please paste the logic again here.

```

> We still need to supply the remaining arguments to the `BenchmarkRunner()`, in our case the `random_state`.

Reviewer: I would like a small sentence before this a la "Notice that the parametrization is incomplete, since we did not specify a value for the `random_state` argument".

Then, just name the argument, i.e. "We now need to supply a value for the `random_state` argument of our benchmark."

```python
# run_benchmarks.py
from nnbench import runner

r = runner.BenchmarkRunner()
result = r.run("./benchmarks.py", params={"random_state": 42})
r.report(to='console', result=result)
```

Executing the parametrized benchmark, we get an output similar to this:

```bash
python run_benchmarks.py  

name                                                 value
------------------------------------------------  --------
benchmark_accuracy_n_estimators=10_max_depth=2    0.955556
benchmark_accuracy_n_estimators=50_max_depth=5    0.866667
benchmark_accuracy_n_estimators=100_max_depth=10  0.911111
```

> # `@product` for benchmark scans over parameter configurations

Reviewer: ```suggestion
## `@nnbench.product` for benchmarks over parameter configuration grids
```

> In case we want to run a benchmark scan for all possible parameter combinations we can use the `@product` decorator and supply the different values for each parameter. 

Reviewer: ```suggestion
In case we want to run a benchmark scan for all possible combinations of a set of parameters, we can use the `@nnbench.product` decorator to supply the different values for each parameter. 
```

```python
# benchmarks.py
import nnbench
from training import prepare_data, train_rf, accuracy

@nnbench.product(n_estimators=[10, 50, 100], max_depth=[2, 5, 10])
def benchmark_accuracy_product(n_estimators: int, max_depth: int, random_state: int) -> float:

>     ... #   Our benchmarking logic from above

Reviewer: See above, add the code.

```

> We still provide the `random_state` parameter to the runner like we did with the `@parametrize` decorator.

Reviewer: ```suggestion
We still provide the `random_state` parameter to the runner directly, like we did with the `@nnbench.parametrize` decorator.
```

> # Benchmarks

Reviewer: It should also be the first page in the user guide since it's so important (first entry in the mkdocs.yml user guide list), so I will wait for this to be merged and rebase before checking in the organization best practices guide.
